addons/website_sale_coupon/models/sale_order.py

```diff
@@ -8,6 +8,17 @@
 class SaleOrder(models.Model):
     _inherit = "sale.order"
 
+    def recompute_coupon_lines(self):
+        if request:
+            code = request.httprequest.cookies.get('pending_coupon_code')
+            coupon = request.env['coupon.coupon'].sudo().search([('code', '=', code)], limit=1)
+
+        for order in self:
+            if code and coupon:
+                coupon._apply_on_order(order)
```

Comment:
This was suggested to me by @JKE-be to prevent the loss of the coupon when it becomes invalid again. The coupon is kept in the cookies and added whenever it becomes valid again. I set the cookie to expire when the coupon expires. What do you think about this method?

Reply:
It makes it impossible to remove the coupon :/
Eg you want to apply another one, you can't since you already have one, and if you remove it, it will recompute it and reapply it directly.
I get jke idea but I think not being able to remove a coupon is more critical, isn't it?
Or maybe investigate if this is possible to hook some code when the remove line is called and detect if we delete ON PURPOSE that line (not deleted by recompute)

Reply:
Good point. It is definitely possible to check the coupon in the cookies for removal though. So I'll try that first.

Reply:
No extra "on purpose" check seems to be needed when removing the order lines. I'm guessing `recompute_coupon_lines` removes the coupon from `applied_coupon_ids` before the related order line gets removed. This is not the case when removing the order line directly from the cart.